test/e2e/network/service.go

```diff
@@ -930,6 +930,109 @@ var _ = common.SIGDescribe("Services", func() {
+
+		defer func() {
+			err := cs.CoreV1().Services(ns).Delete(context.TODO(), serviceName, metav1.DeleteOptions{})
+			framework.ExpectNoError(err, "failed to delete service: %s in namespace: %s", serviceName, ns)
+		}()
+
+		svc1port := "svc1"
+		ginkgo.By("creating service " + serviceName + " in namespace " + ns)
+		svc, err := jig.CreateTCPService(func(service *v1.Service) {
+			service.Spec.Ports = []v1.ServicePort{
+				{
+					Name:       "portname1",
+					Port:       80,
+					TargetPort: intstr.FromString(svc1port),
+				},
+			}
+		})
+		framework.ExpectNoError(err)
+		validateEndpointsPortsOrFail(cs, ns, serviceName, portsByPodName{}) // expect no ep when the svc is just created
+
+		names := map[string]bool{}
+		defer func() {
+			for name := range names {
+				err := cs.CoreV1().Pods(ns).Delete(context.TODO(), name, metav1.DeleteOptions{})
+				framework.ExpectNoError(err, "failed to delete pod: %s in namespace: %s", name, ns)
+			}
+		}()
+
+		port1 := 100
+		containerPorts1 := []v1.ContainerPort{
+			{
+				Name:          svc1port,
+				ContainerPort: int32(port1),
+			},
+		}
+		podname1 := "pod1"
+		createPodOrFail(f, ns, podname1, jig.Labels, containerPorts1, "netexec", "--http-port", strconv.Itoa(port1))
+		names[podname1] = true
+		validateEndpointsPortsOrFail(cs, ns, serviceName, portsByPodName{podname1: {port1}}) // expect pod1 is added as ep
+
+		ginkgo.By("Checking if the Service forwards traffic to pods")
+		execPod := e2epod.CreateExecPodOrFail(cs, ns, "execpod", nil)
+		err = jig.CheckServiceReachability(svc, execPod)
+		framework.ExpectNoError(err)
+
+		ginkgo.By("Adding a new port to service " + serviceName)
+		svc2port := "svc2"
+		svc, err = jig.UpdateService(func(s *v1.Service) {
+			s.Spec.Ports = []v1.ServicePort{
+				{
+					Name:       "portname1",
+					Port:       80,
+					TargetPort: intstr.FromString(svc1port),
+				},
+				{
+					Name:       "portname2",
+					Port:       81,
+					TargetPort: intstr.FromString(svc2port),
+				},
+			}
+		})
+		framework.ExpectNoError(err)
+
+		ginkgo.By("Adding a new endpoint to the new port ")
+		port2 := 101
+		containerPorts2 := []v1.ContainerPort{
+			{
+				Name:          svc2port,
+				ContainerPort: int32(port2),
+			},
+		}
+		podname2 := "pod2"
+		createPodOrFail(f, ns, podname2, jig.Labels, containerPorts2, "netexec", "--http-port", strconv.Itoa(port2))
+		names[podname2] = true
+		validateEndpointsPortsOrFail(cs, ns, serviceName, portsByPodName{podname1: {port1}, podname2: {port2}})
+
+		ginkgo.By("Checking if the Service forwards traffic to pods")
+		err = jig.CheckServiceReachability(svc, execPod)
+		framework.ExpectNoError(err)
+
+		ginkgo.By("Deleting a port from service " + serviceName)
+		svc, err = jig.UpdateService(func(s *v1.Service) {
+			s.Spec.Ports = []v1.ServicePort{
+				{
+					Name:       "portname1",
+					Port:       80,
+					TargetPort: intstr.FromString(svc1port),
+				},
+			}
+		})
+		framework.ExpectNoError(err)
+
```

Comment:
should we also check here that port 80 fails?

Reply:
Yes, it's checked in the next two lines. If port 81 still exists, the actual port will be `portsByPodName{podname1: {port1},podname2: {port2}}` which is different from the expectation (`portsByPodName{podname1: {port1}}`), so the `validateEndpointsPortsOrFail()` will fail. Therefore, port 81 shouldn't even exist.